### PR TITLE
Bug 1744984: Add e2e test for kube controller manager lease configmap deletion

### PR DESCRIPTION
**Pull request by damemi**
This is a work in progress, trying to create a simple e2e that deletes the KCM configmap then attempts to create a replicaset to see if KCM recovered from the deletion, as described here: https://bugzilla.redhat.com/show_bug.cgi?id=1744984

### CODE REVIEW

File: test/e2e/operator_test.go [hunk 1]
```diff
@@ -24,3 +26,68 @@ func TestOperatorNamespace(t *testing.T) {
...
+						},
+					},
+				},
+			},
+		},
+	}
+	_, err = kubeClient.AppsV1().ReplicaSets(operatorclient.OperatorNamespace).Create(replicaSet)
+	if err != nil {
+		t.Fatal(err)
+	}
```
Comment (soltysh) [inline test/e2e/operator_test.go]: You're missing a bit where you wait and verify if that replica set actually run and KCM recovered.

Comment (damemi) [reply test/e2e/operator_test.go]: Updated to add a wait loop that looks for pods created with the replicaset labels, that would indicate the KCM has recovered right?

File: test/e2e/operator_test.go [hunk 2]
```diff
@@ -24,3 +28,80 @@ func TestOperatorNamespace(t *testing.T) {
...
+	for _, pod := range pods.Items {
+		err := kubeClient.CoreV1().Pods("openshift-kube-controller-manager").Delete(pod.Name, &metav1.DeleteOptions{})
+		if err != nil {
+			t.Fatal(err)
+		}
+	}
+
+	// Now try to create a replicaset and see that it is successfully created by the controller manager
```
Comment (deads2k) [inline test/e2e/operator_test.go]: wait for the original pods to be gone before creating this.

File: test/e2e/operator_test.go [hunk 3]
```diff
@@ -24,3 +28,80 @@ func TestOperatorNamespace(t *testing.T) {
...
+	}
+
+	// Try to delete the kube controller manager's configmap in kube-system
+	err = kubeClient.CoreV1().ConfigMaps("kube-system").Delete("kube-controller-manager", &metav1.DeleteOptions{})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Get all the currently running KCM pods, and then delete them
+	pods, err := kubeClient.CoreV1().Pods("openshift-kube-controller-manager").List(metav1.ListOptions{})
```
Comment (deads2k) [inline test/e2e/operator_test.go]: this is pointless

File: test/e2e/operator_test.go [hunk 4]
```diff
@@ -24,3 +28,80 @@ func TestOperatorNamespace(t *testing.T) {
...
+							Name:    "test",
+							Image:   "busybox:latest",
+							Command: []string{"/bin/true"},
+						},
+					},
+				},
+			},
+		},
+	}
+	_, err = kubeClient.AppsV1().ReplicaSets(operatorclient.OperatorNamespace).Create(replicaSet)
```
Comment (deads2k) [inline test/e2e/operator_test.go]: because the pods aren't gone, this doesn't matter

File: test/e2e/operator_test.go [hunk 5]
```diff
@@ -24,3 +28,80 @@ func TestOperatorNamespace(t *testing.T) {
...
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Poll to see if pods from replicaset are created which would indicate KCM recovered
+	wait.Poll(time.Second*5, time.Second*300, func() (bool, error) {
```
Comment (deads2k) [inline test/e2e/operator_test.go]: remove this

File: test/e2e/operator_test.go [hunk 6]
```diff
@@ -24,3 +28,80 @@ func TestOperatorNamespace(t *testing.T) {
...
+		pods, err := kubeClient.CoreV1().Pods(operatorclient.OperatorNamespace).List(metav1.ListOptions{LabelSelector: "app=test"})
+		if err != nil {
+			return false, err
+		}
+		if len(pods.Items) > 0 {
+			return true, nil
+		}
+		return false, nil
+	})
+}
```
Comment (deads2k) [inline test/e2e/operator_test.go]: add a check that the configmap gets recreated

File: test/e2e/operator_test.go [hunk 7]
```diff
@@ -24,3 +27,34 @@ func TestOperatorNamespace(t *testing.T) {
...
+	}
+
+	// Try to delete the kube controller manager's configmap in kube-system
+	err = kubeClient.CoreV1().ConfigMaps("kube-system").Delete("kube-controller-manager", &metav1.DeleteOptions{})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Check to see that the configmap then gets recreated
+	wait.Poll(time.Second*5, time.Second*300, func() (bool, error) {
```
Comment (deads2k) [inline test/e2e/operator_test.go]: you forgot to check this error.

File: test/e2e/operator_test.go [hunk 8]
```diff
@@ -24,3 +27,34 @@ func TestOperatorNamespace(t *testing.T) {
...
+		_, err := kubeClient.CoreV1().ConfigMaps("kube-system").Get("kube-controller-manager", metav1.GetOptions{})
+		if err != nil {
```
Comment (deads2k) [inline test/e2e/operator_test.go]:
```go
if errors.IsNotFound(err){
    return false, nil
}
if err != nil{
    return false, err
}
return true, nil
```

**Comment by soltysh** [comment]
/hold
@damemi it looks like the test keeps failing

**Comment by deads2k** [comment]
> /hold
> @damemi it looks like the test keeps failing

hehe.  I know you didn't doubt me :)

**Comment by damemi** [comment]
Looking more into this, it seems we use 2 roles: `system:openshift:leader-election-lock-cluster-policy-controller` in the `openshift-kube-controller-manager` namespace (see [role](https://github.com/openshift/cluster-kube-controller-manager-operator/blob/master/bindata/v4.1.0/kube-controller-manager/leader-election-cluster-policy-controller-role.yaml), [binding](https://github.com/openshift/cluster-kube-controller-manager-operator/blob/master/bindata/v4.1.0/kube-controller-manager/leader-election-cluster-policy-controller-rolebinding.yaml)) and `system::leader-locking-kube-controller-manager` for the `kube-system` namespace. However, the latter [does not include permissions to create configmaps](https://github.com/kubernetes/kubernetes/blob/292bd0efea53a1416a6a7d01620779d2d72fec96/plugin/pkg/auth/authorizer/rbac/bootstrappolicy/namespace_policy.go#L107-L114), so I'm attempting to change our `openshift` role to a `ClusterRole`, and adding a second binding for it in the `kube-system` namespace

**Comment by damemi** [comment]
Spun up a cluster based on this PR, and it doesn't look like the new rolebinding I added in `kube-system` is created... digging more

**Comment by damemi** [comment]
More info: the `kube-system` rolebinding isn't created, but the (now) ClusterRole is. When I manually create the system rolebinding to that clusterrole and `oc delete cm/kube-controller-manager -n kube-system`, I see the configmap almost instantly recreated. So, if I can just figure out how to get that system rolebinding to be automatically created this should fix the bug and pass the test

**Comment by damemi** [comment]
Splitting the rolebindings into 2 separate files got the operator e2e to pass (maybe resourceapply doesn't do well with a list of resources to parse). Squashing and hoping everything goes green again

**Comment by damemi** [comment]
This is failing upgrade test with the following error:
```
ns/openshift-kube-controller-manager-operator deployment/kube-controller-manager-operator Failed to update RoleBinding.rbac.authorization.k8s.io/system:openshift:leader-election-lock-cluster-policy-controller -n openshift-kube-controller-manager: RoleBinding.rbac.authorization.k8s.io "system:openshift:leader-election-lock-cluster-policy-controller" is invalid: roleRef: Invalid value: rbac.RoleRef{APIGroup:"rbac.authorization.k8s.io", Kind:"ClusterRole", Name:"system:openshift:leader-election-lock-cluster-policy-controller"}: cannot change roleRef (798 times)
```
@deads2k do you have any idea why the `roleRef` would be valid for the other tests, but fail on an upgrade?

**Comment by soltysh** [comment]
> @deads2k do you have any idea why the `roleRef` would be valid for the other tests, but fail on an upgrade?

4.3 is where we split the cluster-policy-controller, that's where this problem might be coming from

**Comment by damemi** [comment]
Ok, switching to 2 namespaced `Roles` (as opposed to 1 `ClusterRole`) seems to fix the problem, and all tests pass. Squashing now and this is ready for final review
cc @soltysh @deads2k 
/bugzilla refresh

File: bindata/v4.1.0/kube-controller-manager/leader-election-cluster-policy-controller-role-kube-system.yaml [hunk 9]
```diff
@@ -0,0 +1,22 @@
+# This role is necessary to create leader lock configmap for upgrades 4.2-> 4.3
+# cluster-policy-controller is split from openshift-controller-manager in 4.3
+# leader lock in openshift-controller-manager NamespaceSecurityAllocationController and in ClusterPolicyController
+# cluster-policy-controller container runs in ns openshift-kube-controller-manager static pod
+# The lock, role, and rolebinding can be removed in 4.4
+apiVersion: rbac.authorization.k8s.io/v1
+kind: Role
+metadata:
+  namespace: kube-system
+  name: system:openshift:leader-election-lock-cluster-policy-controller
```
Comment (deads2k) [inline bindata/v4.1.0/kube-controller-manager/leader-election-cluster-policy-controller-role-kube-system.yaml]: this name doesn't match what you're binding to.  This is a general leader election role

File: bindata/v4.1.0/kube-controller-manager/leader-election-cluster-policy-controller-role-kube-system.yaml [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+rules:
+- apiGroups:
+  - ""
+  resources:
+  - configmaps
+  verbs:
+  - get
+  - list
+  - watch
+  - create
```
Comment (deads2k) [inline bindata/v4.1.0/kube-controller-manager/leader-election-cluster-policy-controller-role-kube-system.yaml]: only create needs to be unconstrained by names. The rest can use the stock role, right?

**Comment by damemi** [comment]
@deads2k updated with your feedback, thanks

**Pull request merged**